### PR TITLE
Only override published resolver when the workspace is different

### LINKED ISSUE #10954: Manifests with a resolver field will be published even if the MSRV is less than 1.51
### Problem

On the recent nightly, it seems that cargo package and cargo publish always generate a resolver field, regardless of MSRV or edition.

This makes it impossible to compile on rustc versions such as 1.50, where the resolver field was unstable, even if the MSRV of the published crate is smaller than 1.51.

```
error: failed to parse manifest at `/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/async-channel-1.7.0/Cargo.toml`

Caused by:
  feature `resolver` is required

  this Cargo does not support nightly features, but if you
  switch to nightly channel you can add
  `cargo-features = ["resolver"]` to enable this feature
```

Probably related to https://github.com/rust-lang/cargo/pull/10911
cc @Nemo157

### Steps

To reproduce:
```console
$ cargo new --lib repro 
$ cd repro
$ sed -i 's/edition = "2021"/edition = "2018"/' Cargo.toml
$ cargo package --allow-dirty
$ cat target/package/repro-0.1.0/Cargo.toml | grep resolver
resolver = "1"
```

### Possible Solution(s)

_No response_

### Notes

_No response_

### Version

```text
cargo 1.65.0-nightly (4fd148c47 2022-08-03)
release: 1.65.0-nightly
commit-hash: 4fd148c47e733770c537efac5220744945d572ef
commit-date: 2022-08-03
host: aarch64-apple-darwin
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.79.1 (sys:0.4.55+curl-7.83.1 system ssl:(SecureTransport) LibreSSL/3.3.6)
os: Mac OS 12.5.0 [64-bit]
```

### PR DESCRIPTION
**Pull request by Nemo157**
### What does this PR try to resolve?

Ensures when publishing a package that uses an implicit `resolver = "1"` to maintain an MSRV before the `resolver` key was stabilized the implicitness is retained rather than being turned into an explicit setting.

fixes #10954 (assuming that the workspace and its packages are configured with a consistent resolver)

### CODE REVIEW

File: src/cargo/util/toml/mod.rs [hunk 1]
```diff
@@ -1342,7 +1342,22 @@ impl TomlManifest {
...
+                    .edition
+                    .as_ref()
+                    .and_then(|e| e.as_defined())
+                    .map(|e| Edition::from_str(e))
+                    .unwrap_or(Ok(Edition::Edition2015))
+                    .map(|e| e.default_resolve_behavior())
+            })?;
+        if ws.resolve_behavior() != current_resolver {
+            package.resolver = Some(ws.resolve_behavior().to_manifest());
+        }
```
Comment (epage) [inline src/cargo/util/toml/mod.rs:1360]: Can you add a comment explaining why this is needed?

File: src/cargo/core/workspace.rs [hunk 2]
```diff
@@ -287,16 +287,12 @@ impl<'cfg> Workspace<'cfg> {
...
-                } else {
-                    None
-                }
-            }),
-            MaybePackage::Virtual(vm) => vm.resolve_behavior(),
+            MaybePackage::Package(p) => p
+                .manifest()
+                .resolve_behavior()
+                .unwrap_or_else(|| p.manifest().edition().default_resolve_behavior()),
+            MaybePackage::Virtual(vm) => vm.resolve_behavior().unwrap_or(ResolveBehavior::V1),
```
Comment (epage) [inline src/cargo/core/workspace.rs:294]: nit: please separate refactorings into a different commit from the bug fix to help reviewers more easily understand intent and the change at hand

**Review by epage — COMMENTED** [review]
Thanks for pulling this out!

The nit is not a blocker.  I mostly included it for consideration for future PRs.  I had considered having it pulled out to minimize risk as this will be backported to beta but I figured having it instead helped reduce risk as it ensures both code paths are aligned.

File: src/cargo/util/toml/mod.rs (hunk 1 above)
Comment (epage) [reply src/cargo/util/toml/mod.rs]: Can you expand on the comment about avoiding inserting `resolver` for old `rust-version`s?

**Comment by epage** [comment]
@ehuss I'm assuming `rust-1.64.0` is what I want to backport this fix to?

**Comment by ehuss** [comment]
Yes, rust-1.64.0 is the correct branch, thanks!

**Pull request merged**